### PR TITLE
React optimization + event based processing for tanks, canisters and pipelines

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

I've converted tanks, canisters, and pipelines fully over to reaction based processing. That means they won't process at roundstart, and if they stop being "interesting", they'll stop. Part of this is making update_parents() only get called when gas is actually moved, rather then all the time.

The idea is to force anything that interacts with a tank/canister/pipeline's gasmixture to use return_air rather then directly accessing it. I need to do a second pass on anything that might be caching it, but I don't think many if any things do that.

I've also included react_hybrid in this pr, which unlike react() gets more expensive with gas "in the mix" count rather then total reaction count. It's cheaper then the old react proc for low gas type mixes, is about parity for 8 gas types, and not too much worse for 21, our max. I've hardcoded the nobilium check too, because doing it in a modular way was slower. If anyone tries to make nobilium 2.0 they can suck it.

Basically, instead of giving each reaction its own priority, we split them into 4 priority groups, pre formation, formation, post formation, and fire. they should just about cover everything you'd want to do with a reaction. This also means you can't return `STOP_REACTIONS` to stop all reactions normally, but I'm fine with that. 

Outside of that I've done some simple optimizations to both react() procs, including as anythings and the removal of an unused check for energy.

Have some profiles from a local:
"Best" case scenario, so 2 gas types, o2 and n2, and no actual reactions being tested or performed.
image

"Middling" case, so a plasma + co2 + n2o + o2 fire started by opening one plasma and one o2 canister in meta toxins, waiting for them to settle, then igniting them with the apc's overload button. Profiling continued until the floor melted away and all gas dissipated.
image

"Worst" case, so all 21 gases in one mixture. (This was without the nobilium hardcode, but it should be close to trivial)
image

These profiles were done on a local with tank and canister reactions disabled after running the following sqdl commands, and waiting for things to equalize. Oh and react_kills_prioirty is what it says on the tin, it just ignored all forms of priority outside of a nobilium check, was included in those tests as a best case senario
```
DELETE /mob/living ///Mobs can open doors, and monkeys breath the bastards
DELETE /obj/machinery/atmospherics ///For obvious reasons
```

I'm looking to test both systems on master with all its havoc and gas diffs, I've already tested the reaction change on master, but there's too many useless reacts to get much good data from that, so let's try em both at once and see what blows up yeah?

There's a potential hole I'm opening up here by the way, if people figure out how to force reactions to only fire when they want them to rather then all the time, we'll have a new sort of bomb. If that happens it should be reasonably easy to resolve, but I figured I'd let you know.

### Current Benefits
Makes the best case react() call ~3x less expensive, while unshackling us from a limited reaction count.
Brings the default load from pipeline processing down from 24 to 16, on meta.
Make react() calls more sane, reduces the amount of unneeded pipeline equalization.
Sets me up for more work.
 
### Some further improvements

We could in theory remove the fire reaction group, but that makes things a bit fiddly
Someone (Me) Should really convert vents and scrubbers to atmos_sensitive, to remove all the unneeded processes they produce. seriously they're a good portion of the 50ms default load atmos has at roundstard

### Post testing
This was close to a micro op, but it's very helpful even outside of that because it removes the weight of caring about reaction length in the absolute sense. Still fucking faster tho.
image

There's a slight cost to this, in that people need to respect the rules around working with tank/canister gas mixtures, I've added a comment to this effect and written a little blurb at the end of Atmospherics.md on the topic

## Why It's Good For The Game

**NO UNNEEDED MOVEMENTS**

## Changelog
:cl: LemonInTheDark, duncathan 
refactor: Gas reactions should be cheaper in the standard case
refactor: Moves canisters, pipelines, and tanks to event based processing, if you see something fail to "update" when it really should let me know yeah?
/:cl:

### CODE REVIEW

File: code/game/objects/items/flamethrower.dm [hunk 1]
```diff
@@ -215,12 +215,14 @@
 /obj/item/flamethrower/proc/default_ignite(turf/target, release_amount = 0.05)
 	//TODO: DEFERRED Consider checking to make sure tank pressure is high enough before doing this...
 	//Transfer 5% of current tank air contents to turf
-	var/datum/gas_mixture/air_transfer = ptank.air_contents.remove_ratio(release_amount)
+	var/datum/gas_mixture/tank_mix = ptank.return_air()
+	var/datum/gas_mixture/air_transfer = tank_mix.remove_ratio(release_amount)
+
 	if(air_transfer.gases[/datum/gas/plasma])
-		air_transfer.gases[/datum/gas/plasma][MOLES] *= 5
+		air_transfer.gases[/datum/gas/plasma][MOLES] *= 5 //Suffering
```
Comment (Ghilker) [inline code/game/objects/items/flamethrower.dm:222]: we should just remove that *= 5

Comment (LemonInTheDark) [reply code/game/objects/items/flamethrower.dm:222]: it's good for gameplay thoooo

**Review by duncathan — APPROVED** [review]
I'm duncathan salt, and I approve this message

**Comment by LemonInTheDark** [comment]
image
it's a micro op, but it also means we don't scale by reaction count anymore, which is quite nice

**Pull request merged**